refactor: Client Classes & Azure OpenAI as a separate Endpoint

danny-avila opened the pull request:
In progress, refactoring to decouple node-chatgpt-api dependency.

This over #527 since I'm taking the time to make a bigger refactor for easier additions/scalability of AI clients.

I will leave the client-refactor branch up for those that want to test localAi support, since it's not fully supported with this refactor yet.

danny-avila committed:
refactor: start new client classes, test localAi support

danny-avila committed:
feat: create base class, extend chatgpt from base

danny-avila committed:
refactor(BaseClient.js): change userId parameter to user
refactor(BaseClient.js): change userId parameter to user
feat(OpenAIClient.js): add sendMessage method
refactor(OpenAIClient.js): change getConversation method to use user parameter instead of userId
refactor(OpenAIClient.js): change saveMessageToDatabase method to use user parameter instead of userId
refactor(OpenAIClient.js): change buildPrompt method to use messages parameter instead of orderedMessages
feat(index.js): export client classes
refactor(askGPTPlugins.js): use req.body.token or process.env.OPENAI_API_KEY as OpenAI API key
refactor(index.js): comment out askOpenAI route
feat(index.js): add openAI route

feat(openAI.js): add new route for OpenAI API requests with support for progress updates and aborting requests.

danny-avila committed:
refactor(BaseClient.js): use optional chaining operator to access messageId property
refactor(OpenAIClient.js): use orderedMessages instead of messages to build prompt
refactor(OpenAIClient.js): use optional chaining operator to access messageId property
refactor(fetch-polyfill.js): remove fetch polyfill
refactor(openAI.js): comment out debug option in clientOptions

danny-avila committed:
refactor: update import statements and remove unused imports in several files
feat: add getAzureCredentials function to azureUtils module
docs: update comments in azureUtils module

danny-avila committed:
refactor(utils): rename migrateConversations to migrateDataToFirstUser for clarity and consistency

danny-avila committed:
feat(chatgpt-client.js): add getAzureCredentials function to retrieve Azure credentials
feat(chatgpt-client.js): use getAzureCredentials function to generate reverseProxyUrl
feat(OpenAIClient.js): add isChatCompletion property to determine if chat completion model is used
feat(OpenAIClient.js): add saveOptions parameter to sendMessage and buildPrompt methods
feat(OpenAIClient.js): modify buildPrompt method to handle chat completion model
feat(openAI.js): modify endpointOption to include modelOptions instead of individual options
refactor(OpenAIClient.js): modify getDelta property to use isChatCompletion property instead of isChatGptModel property
refactor(OpenAIClient.js): modify sendMessage method to use saveOptions parameter instead of modelOptions parameter
refactor(OpenAIClient.js): modify buildPrompt method to use saveOptions parameter instead of modelOptions parameter
refactor(OpenAIClient.js): modify ask method to include endpointOption parameter

danny-avila committed:
chore: delete draft file

danny-avila committed:
refactor(OpenAIClient.js): extract sendCompletion method from sendMessage method for reusability

danny-avila committed:
refactor(BaseClient.js): move sendMessage method to BaseClient class
feat(OpenAIClient.js): inherit from BaseClient class and implement necessary methods and properties for OpenAIClient class.

danny-avila committed:
refactor(BaseClient.js): rename getBuildPromptOptions to getBuildMessagesOptions
feat(BaseClient.js): add buildMessages method to BaseClient class
fix(ChatGPTClient.js): use message.text instead of message.message
refactor(ChatGPTClient.js): rename buildPromptBody to buildMessagesBody
refactor(ChatGPTClient.js): remove console.debug statement and add debug log for prompt variable

refactor(OpenAIClient.js): move setOptions method to the bottom of the class
feat(OpenAIClient.js): add support for cl100k_base encoding
feat(OpenAIClient.js): add support for unofficial chat GPT models
feat(OpenAIClient.js): add support for custom modelOptions
feat(OpenAIClient.js): add caching for tokenizers
feat(OpenAIClient.js): add freeAndInitializeEncoder method to free and reinitialize tokenizers
refactor(OpenAIClient.js): rename getBuildPromptOptions to getBuildMessagesOptions
refactor(OpenAIClient.js): rename buildPrompt to buildMessages
refactor(OpenAIClient.js): remove endpointOption from ask function arguments in openAI.js

danny-avila committed:
refactor(ChatGPTClient.js, OpenAIClient.js): improve code readability and consistency

- In ChatGPTClient.js, update the roleLabel and messageString variables to handle cases where the message object does not have an isCreatedByUser property or a role property with a value of 'user'.
- In OpenAIClient.js, rename the freeAndInitializeEncoder method to freeAndResetEncoder to better reflect its functionality. Also, update the method calls to reflect the new name. Additionally, update the getTokenCount method to handle errors by calling the freeAndResetEncoder method instead of the now-renamed freeAndInitializeEncoder method.

danny-avila committed:
refactor(OpenAIClient.js): extract instructions object to a separate variable and add it to payload after formatted messages
fix(OpenAIClient.js): handle cases where progressMessage.choices is undefined or empty

danny-avila committed:
refactor(BaseClient.js): extract addInstructions method from sendMessage method
feat(OpenAIClient.js): add maxTokensMap object to map maximum tokens for each model
refactor(OpenAIClient.js): use addInstructions method in buildMessages method instead of manually building the payload list

danny-avila committed:
refactor(OpenAIClient.js): remove unnecessary condition for modelOptions.model property in buildMessages method

danny-avila committed:
feat(BaseClient.js): add support for token count tracking and context strategy
feat(OpenAIClient.js): add support for token count tracking and context strategy
feat(Message.js): add tokenCount field to Message schema and updateMessage function

danny-avila committed:
refactor(BaseClient.js): add support for refining messages based on token limit
feat(OpenAIClient.js): add support for context refinement strategy
refactor(OpenAIClient.js): use context refinement strategy in message sending
refactor(server/index.js): improve code readability by breaking long lines

danny-avila committed:
refactor(BaseClient.js): change `remainingContext` to `remainingContextTokens` for clarity
feat(BaseClient.js): add `refinePrompt` and `refinePromptTemplate` to handle message refinement
feat(BaseClient.js): add `refineMessages` method to refine messages
feat(BaseClient.js): add `handleContextStrategy` method to handle context strategy
feat(OpenAIClient.js): add `abortController` to `buildPrompt` method options
refactor(OpenAIClient.js): change `payload` and `tokenCountMap` to let variables in `handleContextStrategy` method
refactor(BaseClient.js): change `remainingContext` to `remainingContextTokens` in `handleContextStrategy` method for consistency
refactor(BaseClient.js): change `remainingContext` to `remainingContextTokens` in `getMessagesWithinTokenLimit` method for consistency
refactor(BaseClient.js): change `remainingContext` to `remainingContext

danny-avila committed:
chore(openAI.js): comment out contextStrategy option in clientOptions

danny-avila committed:
chore(openAI.js): comment out debug option in clientOptions object

danny-avila committed:
test: BaseClient tests in progress

danny-avila committed:
test: Complete OpenAIClient & BaseClient tests

danny-avila committed:
fix(OpenAIClient.js): remove unnecessary whitespace
fix(OpenAIClient.js): remove unused variables and comments
fix(OpenAIClient.test.js): combine getTokenCount and freeAndResetEncoder tests

danny-avila committed:
chore(.eslintrc.js): add rule for maximum of 1 empty line
feat(ask/openAI.js): add abortMessage utility function
fix(ask/openAI.js): handle error and abort message if partial text is less than 2 characters
feat(utils/index.js): export abortMessage utility function

danny-avila committed:
test: complete additional tests

ClaraLeigh commented on api/app/clients/classes/BaseClient.js:
```diff
@@ -0,0 +1,467 @@
+    });
+  }
+
+  setOptions() {
+    throw new Error("Method 'setOptions' must be implemented.");
+  }
+
+  getCompletion() {
+    throw new Error("Method 'getCompletion' must be implemented.");
+  }
+
+  getSaveOptions() {
+    throw new Error('Subclasses must implement getSaveOptions');
+  }
+
+  async buildMessages() {
+    throw new Error('Subclasses must implement buildMessages');
+  }
+
+  getBuildMessagesOptions() {
+    throw new Error('Subclasses must implement getBuildMessagesOptions');
+  }
```
Just thinking here, we might be best creating some error classes for the different error types in the project

ClaraLeigh replied:
although I get that these are yet to be implemented things

ClaraLeigh reviewed (commented):
looking good, only comments is that we could probably refactor some duplicate stuff, move the test classes etc but I'm sure that's already on your mind

ClaraLeigh commented:
oh one more thing, if you get a chance while you're in here. It'd be cool to be able to have the frame work for defining your own STOP, I often want to use END_CODE for a stop point for example

danny-avila committed:
feat(FakeClient.js): add a new class `FakeClient` that extends `BaseClient` and implements methods for a fake client
feat(FakeClient.js): implement the `setOptions` method to handle options for the fake client
feat(FakeClient.js): implement the `initializeFakeClient` function to initialize a fake client with options and fake messages
fix(OpenAIClient.js): remove duplicate `maxTokensMap` import and use the one from utils
feat(BaseClient): return promptTokens and completionTokens

fuegovic commented:
The small issues I previously had are now fixed 👍

danny-avila commented:
> oh one more thing, if you get a chance while you're in here. It'd be cool to be able to have the framework for defining your own STOP, I often want to use END_CODE for a stop point for example

On this, I will do this in a separate PR right after this one. I want to get this functionality in first

danny-avila committed:
feat: Azure OpenAI as a separate endpoint

danny-avila committed:
chore: remove extraneous console logs

danny-avila committed:
fix(azureOpenAI): use chatCompletion endpoint

danny-avila committed:
chore(initializeClient.js): delete initializeClient.js file

chore(askOpenAI.js): delete old OpenAI route handler

chore(handlers.js): remove trailing whitespace in thought variable assignment

danny-avila commented:
closes #548

danny-avila committed:
chore(chatgpt-client.js): remove unused chatgpt-client.js file
refactor(index.js): remove askClient import and export from index.js

danny-avila committed:
chore(chatgpt-client.tokens.js): update test script for memory usage and encoding performance

The test script in `chatgpt-client.tokens.js` has been updated to measure the memory usage and encoding performance of the client. The script now includes information about the initial memory usage, peak memory usage, final memory usage, and memory usage after a timeout. It also provides insights into the number of encoding requests that can be processed per second.

The script has been modified to use the `OpenAIClient` class instead of the `ChatGPTClient` class. Additionally, the number of iterations for the encoding loop has been reduced to 10,000.

A timeout function has been added to simulate a delay of 15 seconds. After the timeout, the memory usage is measured again.

The script now handles uncaught exceptions and logs any errors that occur, except for errors related to failed fetch requests.

Note: This is a test script and should not be used in production

danny-avila committed:
refactor(gptPlugins): refactor ChatAgent to PluginsClient, which extends OpenAIClient

danny-avila committed:
refactor: client paths

danny-avila committed:
chore(jest.config.js): remove jest.config.js file

danny-avila committed:
fix(PluginController.js): update file path to manifest.json
feat(gptPlugins.js): add support for aborting messages

refactor(ask/index.js): rename askGPTPlugins to gptPlugins for consistency

danny-avila committed:
fix(BaseClient.js): fix spacing in generateTextStream function signature
refactor(BaseClient.js): remove unnecessary push to currentMessages in generateUserMessage function
refactor(BaseClient.js): remove unnecessary push to currentMessages in handleStartMethods function
refactor(PluginsClient.js): remove unused variables and date formatting in constructor
refactor(PluginsClient.js): simplify mapping of pastMessages in getCompletionPayload function

danny-avila committed:
refactor(GoogleClient): GoogleClient now extends BaseClient

danny-avila committed:
chore(.env.example): add AZURE_OPENAI_MODELS variable
fix(api/routes/ask/gptPlugins.js): enable Azure integration if PLUGINS_USE_AZURE is true
fix(api/routes/endpoints.js): getOpenAIModels function now accepts options, use AZURE_OPENAI_MODELS if PLUGINS_USE_AZURE is true
fix(client/components/Endpoints/OpenAI/Settings.jsx): remove console.log statement
docs(features/azure.md): add documentation for Azure OpenAI integration and environment variables

danny-avila committed:
fix(e2e:popup): includes the icon + endpoint names in role, name property
